usr.bin/units/units.1
.\" $FreeBSD: head/usr.bin/units/units.1 233510 2012-03-26 15:18:14Z joel $
.Dd April 14, 2014
.Dt UNITS 1
.Os
.Sh NAME
.Nm units
.Nd conversion program
.Sh SYNOPSIS
.Nm
.Op Fl f Ar filename
.Op Fl qvUV
.Op Ar from-unit to-unit
.Sh OPTIONS
The following options are available:
.Bl -tag -width indent
.It Fl f Ar filename
Specify the name of the units data file to load.
.It Fl q
Suppress prompting of the user for units and the display of statistics
about the number of units loaded.
.It Fl U
If the default unit file exists prints its location.  If not, print
.Qo
Units data file not found
.Qc
.It Fl V
Print the version number, usage, and then exit.
.It Ar from-unit to-unit
Allow a single unit conversion to be done directly from the command
line.
The program will not print prompts.
It will print out the
result of the single specified conversion.
.It Fl v
Print the units in the conversion output.  Be more verbose in general.
.El
.Sh DESCRIPTION
The
.Nm
program converts quantities expressed in various scales to
their equivalents in other scales.
The
.Nm
program can only
handle multiplicative or affine scale changes.
It works interactively by prompting
the user for input:
.Bd -literal
    You have: meters
    You want: feet
            * 3.2808399
            / 0.3048

    You have: cm^3
    You want: gallons
            * 0.00026417205
            / 3785.4118

    You have: meters/s
    You want: furlongs/fortnight
            * 6012.8848
            / 0.00016630952

    You have: 1|2 inch
    You want: cm
            * 1.27
            / 0.78740157

    You have: 85 degF
    You want: degC
	29.444444
.Ed
.Pp
Powers of units can be specified using the '^' character as shown in
the example, or by simple concatenation: 'cm3' is equivalent to 'cm^3'.
Multiplication of units can be specified by using spaces, a dash or
an asterisk.
Division of units is indicated by the slash ('/').
Note that multiplication has a higher precedence than division,
so 'm/s/s' is the same as 'm/s^2' or 'm/s s'.
Division of numbers
must be indicated using the vertical bar ('|').
To convert half a
meter, you would write '1|2 meter'.
If you write '1/2 meter' then the
units program would interpret that as equivalent to '0.5/meter'.
If you enter incompatible unit types, the units program will
print a message indicating that the units are not conformable and
it will display the reduced form for each unit:
.Bd -literal
    You have: ergs/hour
    You want: fathoms kg^2 / day
    conformability error
            2.7777778e-11 kg m^2 / sec^3
            2.1166667e-05 kg^2 m / sec
.Ed
.Pp
The conversion information is read from a units data file.
The default
file includes definitions for most familiar units, abbreviations and
metric prefixes.
Some constants of nature included are:
.Pp
.Bl -column -offset indent -compact "mercury"
.It "pi	ratio of circumference to diameter"
.It "c	speed of light"
.It "e	charge on an electron"
.It "g	acceleration of gravity"
.It "force	same as g"
.It "mole	Avogadro's number"
.It "water	pressure per unit height of water"
.It "mercury	pressure per unit height of mercury"
.It "au	astronomical unit"
.El
.Pp
The unit 'pound' is a unit of mass.
Compound names are run together
so 'pound force' is a unit of force.
The unit 'ounce' is also a unit
of mass.
The fluid ounce is 'floz'.
British units that differ from
their US counterparts are prefixed with 'br', and currency is prefixed
with its country name: 'belgiumfranc', 'britainpound'.
When searching
for a unit, if the specified string does not appear exactly as a unit
name, then
.Nm
will try to remove a trailing 's' or a
trailing 'es' and check again for a match.
.Pp
To find out what units are available read the standard units file.
If you want to add your own units you can supply your own file.
A unit is specified on a single line by
giving its name and an equivalence.
Be careful to define
new units in terms of old ones so that a reduction leads to the
primitive units which are marked with '!' characters.
The
.Nm
program will not detect infinite loops that could be caused
by careless unit definitions.
Comments in the unit definition file
begin with a '/' character at the beginning of a line.
.Pp
Prefixes are defined in the same was as standard units, but with
a trailing dash at the end of the prefix name.
If a unit is not found
even after removing trailing 's' or 'es', then it will be checked
against the list of prefixes.
Prefixes will be removed until a legal
base unit is identified.
.Pp
Here is an example of a short units file that defines some basic
units.
.Pp
.Bl -column -offset indent -compact "minute"
.It "m	!a!"
.It "sec	!b!"
.It "micro-	1e-6"
.It "minute	60 sec"
.It "hour	60 min"
.It "inch	0.0254 m"
.It "ft	12 inches"
.It "mile	5280 ft"
.El
.Sh FILES
.Bl -tag -width /usr/share/misc/units.lib -compact
.It Pa /usr/share/misc/units.lib
the standard units library
.El
.Sh AUTHORS
.An Adrian Mariano Aq adrian@cam.cornell.edu
.Sh BUGS
The effect of including a '/' in a prefix is surprising.
.Pp
Exponents entered by the user can be only one digit.
You can work around this by multiplying several terms.
.Pp
The user must use | to indicate division of numbers and / to
indicate division of symbols.
This distinction should not
be necessary.
.Pp
The program contains various arbitrary limits on the length
of the units converted and on the length of the data file.
.Pp
The program should use a hash table to store units so that
it does not take so long to load the units list and check
for duplication.


usr.bin/units/units.c
/*
 * units.c   Copyright (c) 1993 by Adrian Mariano (adrian@cam.cornell.edu)
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 * Disclaimer:  This software is provided by the author "as is".  The author
 * shall not be liable for any damages caused in any way by this software.
 *
 * I would appreciate (though I do not require) receiving a copy of any
 * improvements you might make to this program.
 */

#ifndef lint
static const char rcsid[] =
  "$FreeBSD: head/usr.bin/units/units.c 227194 2011-11-06 08:18:17Z ed $";
#endif /* not lint */

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <histedit.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <sys/capsicum.h>

#include "pathnames.h"

#ifndef UNITSFILE
#define UNITSFILE _PATH_UNITSLIB
#endif

#define MAXUNITS 1000
#define MAXPREFIXES 100

#define MAXSUBUNITS 500

#define PRIMITIVECHAR '!'

static const char *powerstring = "^";

static struct {
	char *uname;
	char *uval;
}      unittable[MAXUNITS];

struct unittype {
	char *numerator[MAXSUBUNITS];
	char *denominator[MAXSUBUNITS];
	double factor;
	double offset;
	int quantity;
};

static struct {
	char *prefixname;
	char *prefixval;
}      prefixtable[MAXPREFIXES];


static char NULLUNIT[] = "";

#ifdef MSDOS
#define SEPARATOR      ";"
#else
#define SEPARATOR      ":"
#endif

static int unitcount;
static int prefixcount;
static bool verbose = false;
static const char * havestr;
static const char * wantstr;


static int	 addsubunit(char *product[], char *toadd);
static int	 addunit(struct unittype *theunit, const char *toadd, int flip, int quantity);
static void	 cancelunit(struct unittype * theunit);
static int	 compare(const void *item1, const void *item2);
static int	 compareproducts(char **one, char **two);
static int	 compareunits(struct unittype * first, struct unittype * second);
static int	 completereduce(struct unittype * unit);
static char	*dupstr(const char *str);
static void	 initializeunit(struct unittype * theunit);
static char	*lookupunit(const char *unit);
static void	 readunits(const char *userfile);
static int	 reduceproduct(struct unittype * theunit, int flip);
static int	 reduceunit(struct unittype * theunit);
static void	 showanswer(struct unittype * have, struct unittype * want);
static void	 showunit(struct unittype * theunit);
static void	 sortunit(struct unittype * theunit);
static void	 usage(void);
static void	 zeroerror(void);

static const char* promptstr = "";

static const char * prompt(EditLine *e __unused) {
	return promptstr;
}

char *
dupstr(const char *str)
{
	char *ret;

	ret = malloc(strlen(str) + 1);
	if (!ret)
		errx(3, "memory allocation error");
	strcpy(ret, str);
	return (ret);
}


void 
readunits(const char *userfile)
{
	FILE *unitfile;
	char line[512], *lineptr;
	int len, linenum, i;
	cap_rights_t unitfilerights;

	unitcount = 0;
	linenum = 0;

	if (userfile) {
		unitfile = fopen(userfile, "rt");
		if (!unitfile)
			errx(1, "unable to open units file '%s'", userfile);
	}
	else {
		unitfile = fopen(UNITSFILE, "rt");
		if (!unitfile) {
			char *direc, *env;
			char filename[1000];

			env = getenv("PATH");
			if (env) {
				direc = strtok(env, SEPARATOR);
				while (direc) {
					snprintf(filename, sizeof(filename),
					    "%s/%s", direc, UNITSFILE);
					unitfile = fopen(filename, "rt");
					if (unitfile)
						break;
					direc = strtok(NULL, SEPARATOR);
				}
			}
			if (!unitfile)
				errx(1, "can't find units file '%s'", UNITSFILE);
		}
	}
	cap_rights_init(&unitfilerights, CAP_READ, CAP_FSTAT);
	if (cap_rights_limit(fileno(unitfile), &unitfilerights) < 0
		&& errno != ENOSYS)
		err(1, "cap_rights_limit() failed");
	while (!feof(unitfile)) {
		if (!fgets(line, sizeof(line), unitfile))
			break;
		linenum++;
		lineptr = line;
		if (*lineptr == '/')
			continue;
		lineptr += strspn(lineptr, " \n\t");
		len = strcspn(lineptr, " \n\t");
		lineptr[len] = 0;
		if (!strlen(lineptr))
			continue;
		if (lineptr[strlen(lineptr) - 1] == '-') { /* it's a prefix */
			if (prefixcount == MAXPREFIXES) {
				warnx("memory for prefixes exceeded in line %d", linenum);
				continue;
			}
			lineptr[strlen(lineptr) - 1] = 0;
			prefixtable[prefixcount].prefixname = dupstr(lineptr);
			for (i = 0; i < prefixcount; i++)
				if (!strcmp(prefixtable[i].prefixname, lineptr)) {
					warnx("redefinition of prefix '%s' on line %d ignored",
					    lineptr, linenum);
					continue;
				}
			lineptr += len + 1;
			lineptr += strspn(lineptr, " \n\t");
			len = strcspn(lineptr, "\n\t");
			if (len == 0) {
				warnx("unexpected end of prefix on line %d",
				    linenum);
				continue;
			}
			lineptr[len] = 0;
			prefixtable[prefixcount++].prefixval = dupstr(lineptr);
		}
		else {		/* it's not a prefix */
			if (unitcount == MAXUNITS) {
				warnx("memory for units exceeded in line %d", linenum);
				continue;
			}
			unittable[unitcount].uname = dupstr(lineptr);
			for (i = 0; i < unitcount; i++)
				if (!strcmp(unittable[i].uname, lineptr)) {
					warnx("redefinition of unit '%s' on line %d ignored",
					    lineptr, linenum);
					continue;
				}
			lineptr += len + 1;
			lineptr += strspn(lineptr, " \n\t");
			if (!strlen(lineptr)) {
				warnx("unexpected end of unit on line %d",
				    linenum);
				continue;
			}
			len = strcspn(lineptr, "\n\t");
			lineptr[len] = 0;
			unittable[unitcount++].uval = dupstr(lineptr);
		}
	}
	fclose(unitfile);
}

void 
initializeunit(struct unittype * theunit)
{
	theunit->numerator[0] = theunit->denominator[0] = NULL;
	theunit->factor = 1.0;
	theunit->offset = 0.0;
	theunit->quantity = 0;
}


int 
addsubunit(char *product[], char *toadd)
{
	char **ptr;

	for (ptr = product; *ptr && *ptr != NULLUNIT; ptr++);
	if (ptr >= product + MAXSUBUNITS) {
		warnx("memory overflow in unit reduction");
		return 1;
	}
	if (!*ptr)
		*(ptr + 1) = NULL;
	*ptr = dupstr(toadd);
	return 0;
}


void 
showunit(struct unittype * theunit)
{
	char **ptr;
	int printedslash;
	int counter = 1;

	printf("%.8g", theunit->factor);
	if (theunit->offset)
		printf("&%.8g", theunit->offset);
	for (ptr = theunit->numerator; *ptr; ptr++) {
		if (ptr > theunit->numerator && **ptr &&
		    !strcmp(*ptr, *(ptr - 1)))
			counter++;
		else {
			if (counter > 1)
				printf("%s%d", powerstring, counter);
			if (**ptr)
				printf(" %s", *ptr);
			counter = 1;
		}
	}
	if (counter > 1)
		printf("%s%d", powerstring, counter);
	counter = 1;
	printedslash = 0;
	for (ptr = theunit->denominator; *ptr; ptr++) {
		if (ptr > theunit->denominator && **ptr &&
		    !strcmp(*ptr, *(ptr - 1)))
			counter++;
		else {
			if (counter > 1)
				printf("%s%d", powerstring, counter);
			if (**ptr) {
				if (!printedslash)
					printf(" /");
				printedslash = 1;
				printf(" %s", *ptr);
			}
			counter = 1;
		}
	}
	if ( counter > 1)
		printf("%s%d", powerstring, counter);
	printf("\n");
}


void 
zeroerror(void)
{
	warnx("unit reduces to zero");
}

/*
   Adds the specified string to the unit.
   Flip is 0 for adding normally, 1 for adding reciprocal.
   Quantity is 1 if this is a quantity to be converted rather than a pure unit.

   Returns 0 for successful addition, nonzero on error.
*/

int 
addunit(struct unittype * theunit, const char *toadd, int flip, int quantity)
{
	char *scratch, *savescr;
	char *item;
	char *divider, *slash, *offset;
	int doingtop;

	if (!strlen(toadd))
		return 1;
	
	savescr = scratch = dupstr(toadd);
	for (slash = scratch + 1; *slash; slash++)
		if (*slash == '-' &&
		    (tolower(*(slash - 1)) != 'e' ||
		    !strchr(".0123456789", *(slash + 1))))
			*slash = ' ';
	slash = strchr(scratch, '/');
	if (slash)
		*slash = 0;
	doingtop = 1;
	do {
		item = strtok(scratch, " *\t\n/");
		while (item) {
			if (strchr("0123456789.", *item)) { /* item is a number */
				double num, offsetnum;

				if (quantity)
					theunit->quantity = 1;

				offset = strchr(item, '&');
				if (offset) {
					*offset = 0;
					offsetnum = atof(offset+1);
				} else
					offsetnum = 0.0;

				divider = strchr(item, '|');
				if (divider) {
					*divider = 0;
					num = atof(item);
					if (!num) {
						zeroerror();
						return 1;
					}
					if (doingtop ^ flip) {
						theunit->factor *= num;
						theunit->offset *= num;
					} else {
						theunit->factor /= num;
						theunit->offset /= num;
					}
					num = atof(divider + 1);
					if (!num) {
						zeroerror();
						return 1;
					}
					if (doingtop ^ flip) {
						theunit->factor /= num;
						theunit->offset /= num;
					} else {
						theunit->factor *= num;
						theunit->offset *= num;
					}
				}
				else {
					num = atof(item);
					if (!num) {
						zeroerror();
						return 1;
					}
					if (doingtop ^ flip) {
						theunit->factor *= num;
						theunit->offset *= num;
					} else {
						theunit->factor /= num;
						theunit->offset /= num;
					}
				}
				if (doingtop ^ flip)
					theunit->offset += offsetnum;
			}
			else {	/* item is not a number */
				int repeat = 1;

				if (strchr("23456789",
				    item[strlen(item) - 1])) {
					repeat = item[strlen(item) - 1] - '0';
					item[strlen(item) - 1] = 0;
				}
				for (; repeat; repeat--)
					if (addsubunit(doingtop ^ flip ? theunit->numerator : theunit->denominator, item))
						return 1;
			}
			item = strtok(NULL, " *\t/\n");
		}
		doingtop--;
		if (slash) {
			scratch = slash + 1;
		}
		else
			doingtop--;
	} while (doingtop >= 0);
	free(savescr);
	return 0;
}


int 
compare(const void *item1, const void *item2)
{
	return strcmp(*(const char * const *)item1, *(const char * const *)item2);
}


void 
sortunit(struct unittype * theunit)
{
	char **ptr;
	unsigned int count;

	for (count = 0, ptr = theunit->numerator; *ptr; ptr++, count++);
	qsort(theunit->numerator, count, sizeof(char *), compare);
	for (count = 0, ptr = theunit->denominator; *ptr; ptr++, count++);
	qsort(theunit->denominator, count, sizeof(char *), compare);
}


void 
cancelunit(struct unittype * theunit)
{
	char **den, **num;
	int comp;

	den = theunit->denominator;
	num = theunit->numerator;

	while (*num && *den) {
		comp = strcmp(*den, *num);
		if (!comp) {
/*      if (*den!=NULLUNIT) free(*den);
      if (*num!=NULLUNIT) free(*num);*/
			*den++ = NULLUNIT;
			*num++ = NULLUNIT;
		}
		else if (comp < 0)
			den++;
		else
			num++;
	}
}




/*
   Looks up the definition for the specified unit.
   Returns a pointer to the definition or a null pointer
   if the specified unit does not appear in the units table.
*/

static char buffer[100];	/* buffer for lookupunit answers with
				   prefixes */

char *
lookupunit(const char *unit)
{
	int i;
	char *copy;

	for (i = 0; i < unitcount; i++) {
		if (!strcmp(unittable[i].uname, unit))
			return unittable[i].uval;
	}

	if (unit[strlen(unit) - 1] == '^') {
		copy = dupstr(unit);
		copy[strlen(copy) - 1] = 0;
		for (i = 0; i < unitcount; i++) {
			if (!strcmp(unittable[i].uname, copy)) {
				strlcpy(buffer, copy, sizeof(buffer));
				free(copy);
				return buffer;
			}
		}
		free(copy);
	}
	if (unit[strlen(unit) - 1] == 's') {
		copy = dupstr(unit);
		copy[strlen(copy) - 1] = 0;
		for (i = 0; i < unitcount; i++) {
			if (!strcmp(unittable[i].uname, copy)) {
				strlcpy(buffer, copy, sizeof(buffer));
				free(copy);
				return buffer;
			}
		}
		if (copy[strlen(copy) - 1] == 'e') {
			copy[strlen(copy) - 1] = 0;
			for (i = 0; i < unitcount; i++) {
				if (!strcmp(unittable[i].uname, copy)) {
					strlcpy(buffer, copy, sizeof(buffer));
					free(copy);
					return buffer;
				}
			}
		}
		free(copy);
	}
	for (i = 0; i < prefixcount; i++) {
		size_t len = strlen(prefixtable[i].prefixname);
		if (!strncmp(prefixtable[i].prefixname, unit, len)) {
			if (!strlen(unit + len) || lookupunit(unit + len)) {
				snprintf(buffer, sizeof(buffer), "%s %s",
				    prefixtable[i].prefixval, unit + len);
				return buffer;
			}
		}
	}
	return 0;
}



/*
   reduces a product of symbolic units to primitive units.
   The three low bits are used to return flags:

     bit 0 (1) set on if reductions were performed without error.
     bit 1 (2) set on if no reductions are performed.
     bit 2 (4) set on if an unknown unit is discovered.
*/


#define ERROR 4

int 
reduceproduct(struct unittype * theunit, int flip)
{

	char *toadd;
	char **product;
	int didsomething = 2;

	if (flip)
		product = theunit->denominator;
	else
		product = theunit->numerator;

	for (; *product; product++) {

		for (;;) {
			if (!strlen(*product))
				break;
			toadd = lookupunit(*product);
			if (!toadd) {
				printf("unknown unit '%s'\n", *product);
				return ERROR;
			}
			if (strchr(toadd, PRIMITIVECHAR))
				break;
			didsomething = 1;
			if (*product != NULLUNIT) {
				free(*product);
				*product = NULLUNIT;
			}
			if (addunit(theunit, toadd, flip, 0))
				return ERROR;
		}
	}
	return didsomething;
}


/*
   Reduces numerator and denominator of the specified unit.
   Returns 0 on success, or 1 on unknown unit error.
*/

int 
reduceunit(struct unittype * theunit)
{
	int ret;

	ret = 1;
	while (ret & 1) {
		ret = reduceproduct(theunit, 0) | reduceproduct(theunit, 1);
		if (ret & 4)
			return 1;
	}
	return 0;
}


int 
compareproducts(char **one, char **two)
{
	while (*one || *two) {
		if (!*one && *two != NULLUNIT)
			return 1;
		if (!*two && *one != NULLUNIT)
			return 1;
		if (*one == NULLUNIT)
			one++;
		else if (*two == NULLUNIT)
			two++;
		else if (strcmp(*one, *two))
			return 1;
		else
			one++, two++;
	}
	return 0;
}


/* Return zero if units are compatible, nonzero otherwise */

int 
compareunits(struct unittype * first, struct unittype * second)
{
	return
	compareproducts(first->numerator, second->numerator) ||
	compareproducts(first->denominator, second->denominator);
}


int 
completereduce(struct unittype * unit)
{
	if (reduceunit(unit))
		return 1;
	sortunit(unit);
	cancelunit(unit);
	return 0;
}

void 
showanswer(struct unittype * have, struct unittype * want)
{
	double ans;

	if (compareunits(have, want)) {
		printf("conformability error\n");
		if (verbose)
			printf("\t%s = ", havestr);
		else
			printf("\t");
		showunit(have);
		if (verbose)
			printf("\t%s = ", wantstr);
		else
			printf("\t");
		showunit(want);
	}
	else if (have->offset != want->offset) {
		if (want->quantity)
			printf("WARNING: conversion of non-proportional quantities.\n");
		if (have->quantity)
			printf("\t%.8g\n",
			    (have->factor + have->offset-want->offset)/want->factor);
		else {
			printf("\t (-> x*%.8g %+.8g)\n\t (<- y*%.8g %+.8g)\n",
			    have->factor / want->factor,
			    (have->offset-want->offset)/want->factor,
			    want->factor / have->factor,
			    (want->offset - have->offset)/have->factor);
		}
	}
	else {
		ans = have->factor / want->factor;
		if (verbose)
			printf("\t%s = %.8g * %s\n", havestr, ans, wantstr);
		else
			printf("\t* %.8g\n", ans);

		if (verbose)
			printf("\t%s = (1 / %.8g) * %s\n", havestr, 1/ans,  wantstr);
		else
			printf("\t/ %.8g\n", 1/ans);
	}
}


void 
usage(void)
{
	fprintf(stderr,
		"usage: units [-f unitsfile] [-UVq] [from-unit to-unit]\n");
	exit(3);
}


int
main(int argc, char **argv)
{

	struct unittype have, want;
	int optchar;
	bool quiet;
	bool readfile;
	History *inhistory;
	EditLine *el;
	HistEvent ev;
	int inputsz;

	quiet = false;
	readfile = false;
	while ((optchar = getopt(argc, argv, "fqvUV:")) != -1) {
		switch (optchar) {
		case 'f':
			readfile = true;
			if (strlen(optarg) == 0)
				readunits(NULL);
			else
				readunits(optarg);
			break;
		case 'q':
			quiet = true;
			break;
		case 'v':
			verbose = true;
			break;
		case 'U':
			if (access(UNITSFILE, F_OK) == 0)
				printf("%s\n", UNITSFILE);
			else
				printf("Units data file not found");
			exit(0);
			break;
		case 'V':
			fprintf(stderr, "FreeBSD units\n");
			usage();
			break;
		default:
			usage();
		}
	}

	if (!readfile)
		readunits(NULL);

	inhistory = history_init();
	el = el_init(argv[0], stdin, stdout, stderr);
	el_set(el, EL_PROMPT, &prompt);
	el_set(el, EL_EDITOR, "emacs");
	el_set(el, EL_SIGNAL, 1);
	el_set(el, EL_HIST, history, inhistory);
	el_source(el, NULL);
	history(inhistory, &ev, H_SETSIZE, 800);
	if (inhistory == 0)
		err(1, "Could not initalize history");

	if (cap_enter() < 0 && errno != ENOSYS)
		err(1, "unable to enter capability mode");

	if (optind == argc - 2) {
		havestr = argv[optind];
		wantstr = argv[optind + 1];
		initializeunit(&have);
		addunit(&have, havestr, 0, 1);
		completereduce(&have);
		initializeunit(&want);
		addunit(&want, wantstr, 0, 1);
		completereduce(&want);
		showanswer(&have, &want);
	}
	else {
		if (!quiet)
			printf("%d units, %d prefixes\n", unitcount,
			    prefixcount);
		for (;;) {
			do {
				initializeunit(&have);
				if (!quiet)
					promptstr = "You have: ";
				havestr = el_gets(el, &inputsz);
				if (havestr == NULL)
					exit(0);
				if (inputsz > 0)
					history(inhistory, &ev, H_ENTER,
					havestr);
			} while (addunit(&have, havestr, 0, 1) ||
			    completereduce(&have));
			do {
				initializeunit(&want);
				if (!quiet)
					promptstr = "You want: ";
				wantstr = el_gets(el, &inputsz);
				if (wantstr == NULL)
					exit(0);
				if (inputsz > 0)
					history(inhistory, &ev, H_ENTER,
					wantstr);
			} while (addunit(&want, wantstr, 0, 1) ||
			    completereduce(&want));
			showanswer(&have, &want);
		}
	}

	history_end(inhistory);
	return(0);
}
